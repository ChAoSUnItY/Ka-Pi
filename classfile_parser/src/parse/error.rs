use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
