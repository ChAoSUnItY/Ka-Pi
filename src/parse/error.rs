use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("internal io error")]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match {0} in {1:?} but got opcode {2}{}", if let Some(opcode) = .3 { format!(" (Which is {:?})", opcode)} else { String::new() })]
    MatchOutOfBoundOpcode(&'static str, Vec<Opcode>, u8, Option<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
