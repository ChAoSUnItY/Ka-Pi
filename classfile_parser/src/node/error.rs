use crate::parse::ParseError;
use thiserror::Error;

pub type NodeResResult<T> = Result<T, NodeResError>;

#[derive(Debug, Error)]
pub enum NodeResError {
    #[error(transparent)]
    ParseFail(#[from] ParseError),
    #[error("failed to parse bytes to string in constant Utf8, given bytes: {0:?}")]
    StringParseFail(Box<[u8]>),
    #[error("unknown constant reference #{0}")]
    UnknownConstantReference(u16),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBound(&'static str, Vec<&'static str>, usize),
    #[error("expected referenced constant {0} at #{1} but got {2}")]
    MismatchReferenceConstant(&'static str, u16, &'static str),
    #[error("character {0} cannot be converted into wildcard")]
    InvalidWildcard(char),
    #[error("character {0} cannot be converted into base type")]
    InvalidBaseType(char),
}
