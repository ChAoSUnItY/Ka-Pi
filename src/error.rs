use std::{
    error::Error,
    fmt::{Debug, Display},
};
use std::fmt::write;

use jni::errors::Error as JniError;

use crate::error::KapiError::*;

pub(crate) trait IntoKapiResult<T> {
    fn into_kapi(self) -> KapiResult<T>;
}

pub type KapiResult<T> = Result<T, KapiError>;

impl<T, E> IntoKapiResult<T> for Result<T, E>
where
    E: Into<KapiError>,
{
    fn into_kapi(self) -> KapiResult<T> {
        self.map_err(|e| e.into())
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum KapiError {
    StateError(&'static str),
    Utf8Error(&'static str),
    TypeError(&'static str),
    ArgError(String),
    ClassResolveError(&'static str),
    ClassParseError(String),
    JNIError(String),
}

impl Display for KapiError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StateError(cause) => write!(f, "{}", cause),
            Utf8Error(cause) => write!(f, "{}", cause),
            TypeError(cause) => write!(f, "{}", cause),
            ArgError(cause) => write!(f, "{}", cause),
            ClassResolveError(cause) => write!(f, "{}", cause),
            ClassParseError(cause) => write!(f, "{}", cause),
            JNIError(cause) => write!(f, "{}", cause),
        }
    }
}

impl Error for KapiError {}

impl From<jni::errors::Error> for KapiError {
    fn from(value: JniError) -> Self {
        JNIError(value.to_string())
    }
}
