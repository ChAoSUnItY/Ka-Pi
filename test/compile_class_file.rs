use insta::assert_yaml_snapshot;

use ka_pi::error::KapiResult;

#[test]
fn test_main() -> KapiResult<()> {
    let class_tree = ka_pi::asm::parse::to_class(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    ))?;

    assert_yaml_snapshot!(class_tree);

    Ok(())
}
