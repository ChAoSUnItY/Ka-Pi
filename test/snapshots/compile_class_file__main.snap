---
source: test/compile_class_file.rs
expression: class_tree
---
java_version: V17
constant_pool_count: 34
constant_pool:
  len: 34
  entries:
    1:
      MethodRef:
        class_index: 2
        name_and_type_index: 3
    2:
      Class:
        name_index: 4
    3:
      NameAndType:
        name_index: 5
        type_index: 6
    4:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 79
          - 98
          - 106
          - 101
          - 99
          - 116
    5:
      Utf8:
        length: 6
        bytes:
          - 60
          - 105
          - 110
          - 105
          - 116
          - 62
    6:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 86
    7:
      FieldRef:
        class_index: 8
        name_and_type_index: 9
    8:
      Class:
        name_index: 10
    9:
      NameAndType:
        name_index: 11
        type_index: 12
    10:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 121
          - 115
          - 116
          - 101
          - 109
    11:
      Utf8:
        length: 3
        bytes:
          - 111
          - 117
          - 116
    12:
      Utf8:
        length: 21
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 105
          - 111
          - 47
          - 80
          - 114
          - 105
          - 110
          - 116
          - 83
          - 116
          - 114
          - 101
          - 97
          - 109
          - 59
    13:
      String:
        string_index: 14
    14:
      Utf8:
        length: 12
        bytes:
          - 72
          - 101
          - 108
          - 108
          - 111
          - 32
          - 119
          - 111
          - 114
          - 108
          - 100
          - 33
    15:
      MethodRef:
        class_index: 16
        name_and_type_index: 17
    16:
      Class:
        name_index: 18
    17:
      NameAndType:
        name_index: 19
        type_index: 20
    18:
      Utf8:
        length: 19
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 105
          - 111
          - 47
          - 80
          - 114
          - 105
          - 110
          - 116
          - 83
          - 116
          - 114
          - 101
          - 97
          - 109
    19:
      Utf8:
        length: 7
        bytes:
          - 112
          - 114
          - 105
          - 110
          - 116
          - 108
          - 110
    20:
      Utf8:
        length: 21
        bytes:
          - 40
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
          - 41
          - 86
    21:
      Class:
        name_index: 22
    22:
      Utf8:
        length: 4
        bytes:
          - 77
          - 97
          - 105
          - 110
    23:
      Utf8:
        length: 4
        bytes:
          - 67
          - 111
          - 100
          - 101
    24:
      Utf8:
        length: 15
        bytes:
          - 76
          - 105
          - 110
          - 101
          - 78
          - 117
          - 109
          - 98
          - 101
          - 114
          - 84
          - 97
          - 98
          - 108
          - 101
    25:
      Utf8:
        length: 18
        bytes:
          - 76
          - 111
          - 99
          - 97
          - 108
          - 86
          - 97
          - 114
          - 105
          - 97
          - 98
          - 108
          - 101
          - 84
          - 97
          - 98
          - 108
          - 101
    26:
      Utf8:
        length: 4
        bytes:
          - 116
          - 104
          - 105
          - 115
    27:
      Utf8:
        length: 6
        bytes:
          - 76
          - 77
          - 97
          - 105
          - 110
          - 59
    28:
      Utf8:
        length: 4
        bytes:
          - 109
          - 97
          - 105
          - 110
    29:
      Utf8:
        length: 22
        bytes:
          - 40
          - 91
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
          - 41
          - 86
    30:
      Utf8:
        length: 4
        bytes:
          - 97
          - 114
          - 103
          - 115
    31:
      Utf8:
        length: 19
        bytes:
          - 91
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
    32:
      Utf8:
        length: 10
        bytes:
          - 83
          - 111
          - 117
          - 114
          - 99
          - 101
          - 70
          - 105
          - 108
          - 101
    33:
      Utf8:
        length: 9
        bytes:
          - 77
          - 97
          - 105
          - 110
          - 46
          - 106
          - 97
          - 118
          - 97
access_flags:
  - Public
  - Super
this_class: 21
super_class: 2
interfaces_count: 0
interfaces: []
fields_count: 0
fields: []
methods_count: 2
methods:
  - access_flags:
      - Public
    name_index: 5
    descriptor_index: 6
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 23
        attribute_len: 47
        info:
          - 0
          - 1
          - 0
          - 1
          - 0
          - 0
          - 0
          - 5
          - 42
          - 183
          - 0
          - 1
          - 177
          - 0
          - 0
          - 0
          - 2
          - 0
          - 24
          - 0
          - 0
          - 0
          - 6
          - 0
          - 1
          - 0
          - 0
          - 0
          - 1
          - 0
          - 25
          - 0
          - 0
          - 0
          - 12
          - 0
          - 1
          - 0
          - 0
          - 0
          - 5
          - 0
          - 26
          - 0
          - 27
          - 0
          - 0
        attribute:
          Code:
            max_stack: 1
            max_locals: 1
            code_length: 5
            code:
              - 42
              - 183
              - 0
              - 1
              - 177
            instructions:
              - ALOAD_0
              - INVOKESPECIAL:
                  index: 1
              - RETURN
            exception_table_length: 0
            exception_table: []
            attributes_length: 2
            attributes:
              - attribute_name_index: 24
                attribute_len: 6
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 1
                attribute:
                  LineNumberTable:
                    line_number_table_length: 1
                    line_number_table:
                      - start_pc: 0
                        line_number: 1
              - attribute_name_index: 25
                attribute_len: 12
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 5
                  - 0
                  - 26
                  - 0
                  - 27
                  - 0
                  - 0
                attribute: ~
  - access_flags:
      - Public
      - Static
    name_index: 28
    descriptor_index: 29
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 23
        attribute_len: 55
        info:
          - 0
          - 2
          - 0
          - 1
          - 0
          - 0
          - 0
          - 9
          - 178
          - 0
          - 7
          - 18
          - 13
          - 182
          - 0
          - 15
          - 177
          - 0
          - 0
          - 0
          - 2
          - 0
          - 24
          - 0
          - 0
          - 0
          - 10
          - 0
          - 2
          - 0
          - 0
          - 0
          - 3
          - 0
          - 8
          - 0
          - 4
          - 0
          - 25
          - 0
          - 0
          - 0
          - 12
          - 0
          - 1
          - 0
          - 0
          - 0
          - 9
          - 0
          - 30
          - 0
          - 31
          - 0
          - 0
        attribute:
          Code:
            max_stack: 2
            max_locals: 1
            code_length: 9
            code:
              - 178
              - 0
              - 7
              - 18
              - 13
              - 182
              - 0
              - 15
              - 177
            instructions:
              - GETSTATIC:
                  index: 7
              - LDC:
                  index: 13
              - INVOKEVIRTUAL:
                  index: 15
              - RETURN
            exception_table_length: 0
            exception_table: []
            attributes_length: 2
            attributes:
              - attribute_name_index: 24
                attribute_len: 10
                info:
                  - 0
                  - 2
                  - 0
                  - 0
                  - 0
                  - 3
                  - 0
                  - 8
                  - 0
                  - 4
                attribute:
                  LineNumberTable:
                    line_number_table_length: 2
                    line_number_table:
                      - start_pc: 0
                        line_number: 3
                      - start_pc: 8
                        line_number: 4
              - attribute_name_index: 25
                attribute_len: 12
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 9
                  - 0
                  - 30
                  - 0
                  - 31
                  - 0
                  - 0
                attribute: ~
attributes_count: 1
attributes:
  - attribute_name_index: 32
    attribute_len: 2
    info:
      - 0
      - 33
    attribute:
      SourceFile:
        source_file_index: 33

